### ISSUE TITLE
`vote_stream` is sometimes closed before the tally is met

### ISSUE THREAD

**Issue by youngjoon-lee** [logos-co/nomos-node#329]
Based on PR #326 
When running the `ten_nodes_happy` test with multiple mixnodes in the exit layer, the following error sometimes occur.
```
waiting... 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0
waiting... 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0
waiting... 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0
waiting... 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0
waiting... 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0
waiting... 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0 | 0
...
thread 'Overwatch' panicked at 'internal error: entered unreachable code', /Users/yjlee/repos/nomos-node-mix/nomos-services/consensus/src/tally/happy.rs:83:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread 'ten_nodes_happy' panicked at 'called `Result::unwrap()` on an `Err` value: reqwest::Error { kind: Decode, source: Error("expected value", line: 1, column: 1) }', tests/src/nodes/nomos.rs:228:14
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: test failed, to rerun pass `-p tests --test test_consensus_happy_path`
```
The `entered unreachable code` occurs because the `vote_stream` returned `None` before the tally is met.
This occurs even before the view 0 is approved.

I added a log to see how many votes has been received before the error occurs:
```
nomos_consensus::tally::happy: UNREACHABLE: threshold:10, seen.len():9
```

This doesn't happen if we run the test with only one mixnode in the exit layer (it means the case when all messages are eventually broadcasted from the only one nomos-node).

**Comment by youngjoon-lee** [comment]
> I added a log to see how many votes has been received before the error occurs:
> ```
> nomos_consensus::tally::happy: UNREACHABLE: threshold:10, seen.len():9
> ```

Added more logs and found that the vote from the leader itself isn't contained in the stream.
```
2023-08-25T10:22:41.673832Z ERROR gather_votes{committee=Committee { members: {NodeId([104, 86, 235, 49, 174, 105, 181, 48, 202, 216, 237, 69, 9, 120, 19, 198, 200, 121, 161, 58, 232, 117, 183, 97, 173, 117, 48, 182, 36, 194, 71, 63])} } block=Block { id: BlockId([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), view: View(0), parent_qc: Standard(StandardQc { view: View(-1), id: BlockId([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]) }), leader_proof: LeaderId { leader_id: NodeId([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]) } }}:
nomos_consensus::tally::happy: UNREACHABLE: threshold:10, seen.len():9, seen:{NodeId([120, 239, 89, 239, 95, 95, 226, 46, 32, 166, 206, 111, 79, 250, 186, 118, 221, 20, 225, 232, 29, 130, 152, 37, 180, 198, 85, 116, 155, 12, 37, 2]), NodeId([32, 234, 179, 173, 253, 144, 1, 127, 32, 181, 195, 99, 152, 204, 54, 39, 208, 220, 99, 211, 8, 93, 223, 40, 32, 156, 30, 210, 247, 173, 176, 246]), NodeId([65, 147, 179, 22, 12, 6, 170, 89, 74, 224, 203, 31, 245, 164, 243, 53, 69, 220, 203, 52, 164, 186, 145, 252, 244, 187, 33, 70, 252, 30, 171, 92]), NodeId([228, 38, 222, 2, 23, 103, 202, 126, 131, 253, 33, 199, 232, 197, 158, 216, 244, 126, 171, 39, 68, 246, 175, 15, 187, 232, 71, 199, 75, 74, 236, 121]), NodeId([78, 137, 205, 25, 164, 180, 191, 30, 92, 152, 110, 59, 108, 250, 104, 245, 4, 48, 91, 174, 140, 237, 172, 70, 70, 237, 32, 240, 22, 178, 197, 161]), NodeId([8, 138, 166, 169, 166, 186, 72, 232, 68, 221, 187, 208, 82, 38, 204, 22, 226, 4, 134, 235, 21, 162, 240, 106, 124, 11, 107, 74, 177, 185, 239, 201]), NodeId([181, 71, 24, 176, 172, 201, 28, 131, 139, 151, 164, 81, 65, 27, 231, 101, 227, 204, 34, 47, 226, 252, 244, 188, 32, 142, 195, 190, 107, 143, 77, 59]), NodeId([148, 104, 116, 181, 130, 114, 114, 254, 40, 209, 29, 185, 168, 17, 95, 67, 49, 209, 187, 19, 191, 104, 63, 249, 121, 253, 199, 107, 56, 167, 254, 78]), NodeId([184, 117, 220, 67, 94, 222, 239, 25, 91, 144, 221, 128, 154, 7, 28, 59, 36, 217, 38, 49, 239, 69, 173, 0, 87, 118, 122, 175, 7, 34, 73, 48])}
```

**Comment by danielSanchezQ** [comment]
> > I added a log to see how many votes has been received before the error occurs:
> > ```
> > nomos_consensus::tally::happy: UNREACHABLE: threshold:10, seen.len():9
> > ```
> 
> Added more logs and found that the vote from the leader itself isn't contained in the stream.
> 
> ```
> 2023-08-25T10:22:41.673832Z ERROR gather_votes{committee=Committee { members: {NodeId([104, 86, 235, 49, 174, 105, 181, 48, 202, 216, 237, 69, 9, 120, 19, 198, 200, 121, 161, 58, 232, 117, 183, 97, 173, 117, 48, 182, 36, 194, 71, 63])} } block=Block { id: BlockId([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]), view: View(0), parent_qc: Standard(StandardQc { view: View(-1), id: BlockId([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]) }), leader_proof: LeaderId { leader_id: NodeId([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]) } }}:
> nomos_consensus::tally::happy: UNREACHABLE: threshold:10, seen.len():9, seen:{NodeId([120, 239, 89, 239, 95, 95, 226, 46, 32, 166, 206, 111, 79, 250, 186, 118, 221, 20, 225, 232, 29, 130, 152, 37, 180, 198, 85, 116, 155, 12, 37, 2]), NodeId([32, 234, 179, 173, 253, 144, 1, 127, 32, 181, 195, 99, 152, 204, 54, 39, 208, 220, 99, 211, 8, 93, 223, 40, 32, 156, 30, 210, 247, 173, 176, 246]), NodeId([65, 147, 179, 22, 12, 6, 170, 89, 74, 224, 203, 31, 245, 164, 243, 53, 69, 220, 203, 52, 164, 186, 145, 252, 244, 187, 33, 70, 252, 30, 171, 92]), NodeId([228, 38, 222, 2, 23, 103, 202, 126, 131, 253, 33, 199, 232, 197, 158, 216, 244, 126, 171, 39, 68, 246, 175, 15, 187, 232, 71, 199, 75, 74, 236, 121]), NodeId([78, 137, 205, 25, 164, 180, 191, 30, 92, 152, 110, 59, 108, 250, 104, 245, 4, 48, 91, 174, 140, 237, 172, 70, 70, 237, 32, 240, 22, 178, 197, 161]), NodeId([8, 138, 166, 169, 166, 186, 72, 232, 68, 221, 187, 208, 82, 38, 204, 22, 226, 4, 134, 235, 21, 162, 240, 106, 124, 11, 107, 74, 177, 185, 239, 201]), NodeId([181, 71, 24, 176, 172, 201, 28, 131, 139, 151, 164, 81, 65, 27, 231, 101, 227, 204, 34, 47, 226, 252, 244, 188, 32, 142, 195, 190, 107, 143, 77, 59]), NodeId([148, 104, 116, 181, 130, 114, 114, 254, 40, 209, 29, 185, 168, 17, 95, 67, 49, 209, 187, 19, 191, 104, 63, 249, 121, 253, 199, 107, 56, 167, 254, 78]), NodeId([184, 117, 220, 67, 94, 222, 239, 25, 91, 144, 221, 128, 154, 7, 28, 59, 36, 217, 38, 49, 239, 69, 173, 0, 87, 118, 122, 175, 7, 34, 73, 48])}
> ```

The vote is broadcasted right? If the node is subscribed it should reach.

**Comment by youngjoon-lee** [comment]
> The vote is broadcasted right? If the node is subscribed it should reach. Unless we are filtering self-messages somehow.

Yes, it's broadcasted. I'm testing with adding more logs. I can see the VoteMsg is broadcased from the leader, as intended. But, the msg doesn't arrive in anywhere.
Also, I'm curious how the `Stream` can be closed from the sender side (The unreachable error means that the stream has been closed from the sender).
I'm taking a look.

**Comment by youngjoon-lee** [comment]
I found that this `unreachable` happens only when local_timeout and timeout_qc sometimes occur. The stream is closed from the sender side in this case because the libp2p message cache is advanced for the newer views. 
The current timeout setting is 10s in the integration test. I'm taking a look why timeout happens.

**Comment by youngjoon-lee** [comment]
> I'm still taking a look why enough block approvals aren't gathered within 10s. I guess some messages are dropped from somewhere, but haven't found it yet.

I found the reason. 
```
2023-08-28T06:20:49.857058Z ERROR nomos_network::backends::libp2p: failed to broadcast message to topic: /carnot/proto InsufficientPeers
```
Sometimes, libp2p broadcasting fails with the `InsufficientPeers` error, which happens when the libp2p peer doesn't have any peers subscribed to the topic, in its peer store. After the libp2p node is spawned, it takes some time to get GRAFT messages that say who is subscribing what. 

If broadcasting is started before peer store is populated, the broadcasting is failed. And, we don't retry broadcasting. This situation might be okay in production, but our integration test has more strict setting: `leader_threshold = 100%`.

This error is not very related to the mixnet, but this happen more often when we use mixnet. The message is eventually broadcasted from a random node in the mixnet exit layer. That node most likely hasn't initialized its peer store yet because of the order of spawning nodes. 


We may think about retrying broadcasting when `InsufficientPeers` happens. But, I'm not sure if it hasn't any side effects.

**Comment by danielSanchezQ** [comment]
> > I'm still taking a look why enough block approvals aren't gathered within 10s. I guess some messages are dropped from somewhere, but haven't found it yet.
> 
> I found the reason.
> 
> ```
> 2023-08-28T06:20:49.857058Z ERROR nomos_network::backends::libp2p: failed to broadcast message to topic: /carnot/proto InsufficientPeers
> ```
> 
> Sometimes, libp2p broadcasting fails with the `InsufficientPeers` error, which happens when the libp2p peer doesn't have any peers subscribed to the topic, in its peer store. After the libp2p node is spawned, it takes some time to get GRAFT messages that say who is subscribing what.
> 
> If broadcasting is started before peer store is populated, the broadcasting is failed. And, we don't retry broadcasting. This situation might be okay in production, but our integration test has more strict setting: `leader_threshold = 100%`.
> 
> This error is not very related to the mixnet, but this happen more often when we use mixnet. The message is eventually broadcasted from a random node in the mixnet exit layer. That node most likely hasn't initialized its peer store yet because of the order of spawning nodes.
> 
> We may think about retrying broadcasting when `InsufficientPeers` happens. But, I'm not sure if it hasn't any side effects.

I think in this case the simplest solution is to wait for the nodes to subscribe in the integration tests. 
In production if broadcasting fails, we may try again. But we need to actually analyze what are the cases when broadcasting may fail, as we may not care about them all.

**Comment by youngjoon-lee** [comment]
> I think in this case the simplest solution is to wait for the nodes to subscribe in the integration tests.
In production if broadcasting fails, we may try again. But we need to actually analyze what are the cases when broadcasting may fail, as we may not care about them all.

Right, but the issue is that we aren't able to wait for nodes to subscribe to the topic **at the testing level** :). Once we spawn nomos-nodes, they immediately start broadcasting messages. I'm thinking of adding an option, so that nodes can wait until they have at least one peer at their peer store or they can retry broadcasting for some cases. At the same time, I'm trying to not make this too verbose because this might not be a huge issue in production.

**Comment by danielSanchezQ** [comment]
> > I think in this case the simplest solution is to wait for the nodes to subscribe in the integration tests.
> > In production if broadcasting fails, we may try again. But we need to actually analyze what are the cases when broadcasting may fail, as we may not care about them all.
> 
> Right, but the issue is that we aren't able to wait for nodes to subscribe to the topic **at the testing level** :). Once we spawn nomos-nodes, they immediately start broadcasting messages. I'm thinking of adding an option, so that nodes can wait until they have at least one peer at their peer store or they can retry broadcasting for some cases. At the same time, I'm trying to not make this too verbose because this might not be a huge issue in production.

Most probably we can have a call in the Network service to validate stuff. For example, ping (reply to a channel) me (other sesvice) when this is ready.  And then other service can decide what to do (maybe even stop, or retry after `n` seconds or something). Then we can use it in the adapter to wait for network ready before starting consensus.

**Comment by youngjoon-lee** [comment]
>Most probably we can have a call in the Network service to validate stuff. For example, ping (reply to a channel) me (other sesvice) when this is ready. And then other service can decide what to do (maybe even stop, or retry after n seconds or something). Then we can use it in the adapter to wait for network ready before starting consensus.

Yes, that would be a great idea, having a 'ready-check' concept for services. I think this issue is related to both libp2p and waku, regardless of with/without mixnet. I'll make a new issue for this, if you're fine.

**Comment by Zeegomo** [comment]
I think long term we should:
* allow for vote_stream to be closed early as I don't think we can enforce ordering between task cancellation and stream closing. Ideally we should check that the reason for closing is what we expect (in this case, we changed view). This case is not currently covered by our testcases (not sistematically at least) and our implementation does not support it. 
* retry messages that errored with insufficient peers


a ready-check is somewhat difficult if the network conditions can be altered at runtime (e.g. I receive a mixnet message to be published to a topic I'm not subscribed to)

**Comment by youngjoon-lee** [comment]
I agree. But still it would be great if Overwatch supports the ready-check concept or the order of service initializations, for other purposes. For example, the logs printed before the log service is initialized are not actually printed in the output file or stdout.
For the `vote_stream`, I am happy with the solutions that Zeegomo suggested.

**Comment by Zeegomo** [comment]
Reopening because the issue is not resolved in all cases. We still do not cover the possibility where the stream get closed before the task is canceled

**Comment by danielSanchezQ** [comment]
> Reopening because the issue is not resolved in all cases. We still do not cover the possibility where the stream get closed before the task is canceled

Should the task be cancel if the stream is gone? Meaning that it just couldn't collect the votes, for one reason or another?

**Comment by Zeegomo** [comment]
> Should the task be cancel if the stream is gone?

Yes, and it's what we do unless there are bugs somewhere, the issue is that we can't (i think) reliably control the order in which canceling the task and the stream happen (could be that we cancel the stream and poll the task once more before canceling it). Ultimately, I think we just should not panic if the stream is closed and we didn't complete the tally

**Comment by youngjoon-lee** [comment]
>  Ultimately, I think we just should not panic if the stream is closed and we didn't complete the tally

I agree (https://github.com/logos-co/nomos-node/issues/329#issuecomment-1695046531)

**Comment by danielSanchezQ** [comment]
@youngjoon-lee @zeegomo Can we consider this fixed now?

**Comment by youngjoon-lee** [comment]
> @youngjoon-lee @zeegomo Can we consider this fixed now?

Let me check. IIRC, this issue is not resolved in all cases, but I need to check it again

**Comment by youngjoon-lee** [comment]
> > @youngjoon-lee @zeegomo Can we consider this fixed now?
> 
> Let me check. IIRC, this issue is not resolved in all cases, but I need to check it again

We can consider this issue has been resolved by https://github.com/logos-co/nomos-node/pull/415/commits/8e78539693ad921518d9fcb09ec22ddc6e251d4f. By that, we now return an error instead of panicking (`unreachable!()`) when `vote_stream` is closed before the tally is met. The reason why `vote_stream` is closed is timeout, as described in https://github.com/logos-co/nomos-node/issues/329#issuecomment-1695046531. And, the timeout can happen in various reasons (https://github.com/logos-co/nomos-node/pull/415/commits/f57bfc846ef39f188bc8b679d1012a702b66c696 and https://github.com/logos-co/nomos-node/pull/415/commits/f0698902b491e4de7f5bdfed6116f33a95cab53e in the integration test). Although we don't have a mechanism which guarantess the order between vote_stream close and tally cancellation (as Giacomo mentioned: https://github.com/logos-co/nomos-node/issues/329#issuecomment-1713477404), but I think it's okay for now to just return an error instead of panic.
